core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/variable/listener/VariableListenerWithSources.java

```diff
@@ -0,0 +1,41 @@
+package org.optaplanner.core.impl.domain.variable.listener;
+
+import java.util.Collection;
+import java.util.Collections;
+
+import org.optaplanner.core.api.domain.variable.AbstractVariableListener;
+import org.optaplanner.core.impl.domain.variable.descriptor.VariableDescriptor;
+
+/**
+ * Holds a variable listener and all its source variable descriptors.
+ *
+ * @param <Solution_>
+ */
+public final class VariableListenerWithSources<Solution_> {
+
+    private final AbstractVariableListener<Solution_, Object> variableListener;
+    private final Collection<VariableDescriptor<Solution_>> sourceVariableDescriptors;
```

Comment:
When would there be more than one? Only for the deprecated custom shadow variable annotation code?

Reply:
Not only for `@CustomShadowVariable` but also for repeated `@ShadowVariable` using the same variable listener class. This is typical for chained models (see CSG `BusStop` or `Shuttle`) but also non-chained models (PJS `Allocation`) and list variable models as well (VRP `TimeWindowedCustomer`).